app/models/potential_matcher/prompt_tag_type_info.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+# A class used for gathering information about the rough frequency of different
+# required tag types in all of the prompts in a collection. Used to figure out
+# what would be a good "index type" -- that is, a type that we can use to try
+# to look up matches with.
+class PromptTagTypeInfo
+  # We use a larger batch size here because prompts are smaller than signups
+  # (which can contain many prompts), so we can store more of them in memory at
+  # the same time.
+  def initialize(collection, batch_size = 300)
+    @collection = collection
+    @batch_size = batch_size
+
+    load_collection_info
+    initialize_data_tables
+  end
+
+  private
+
+  # Load information about the passed-in collection.
+  def load_collection_info
+    @total_prompts = @collection.prompts.count
+    @total_signups = @collection.signups.count
+    @settings = @collection.challenge.potential_match_settings
+    @required_types = @settings.required_types
+
+    @limit_first_tags = [10, @total_signups / 10].min
+    @limit_prompts_with_any = [10, @total_prompts / 10].max
+  end
+
+  # Set up @count_prompts_with_any and @first_tags_of_type with the correct
+  # defaults.
+  def initialize_data_tables
+    @count_prompts_with_any = {}
+    @first_tags_of_type = {}
+
+    @required_types.each do |type|
+      @count_prompts_with_any[type] = 0
+      @first_tags_of_type[type] = []
+    end
+  end
+
+  # A good type to use for indexing has a fair number of tags of that type (so
+  # that we aren't, e.g. trying to match prompts on fandom in a single-fandom
+  # exchange), and doesn't have very many people requesting "any" (so that we
+  # have a lot of constraints on who can match with whom).
+  def calculate_good_index_types
+    @good_index_types = @required_types.select do |type|
+      (@first_tags_of_type[type].size >= @limit_first_tags) &&
+        (@count_prompts_with_any[type] <= @limit_prompts_with_any)
```

Comment:
Hmm, interesting. I think part of the problem is that I approached "goodness" as a binary, when really it's a spectrum. If that MCU challenge had a lot of different freeforms, and no one requested Any Freeform, then characters would be a bad tag type to index on, and we wouldn't want it to be included in the set of "good" tag types (since we'd want it to be indexed on freeforms instead). But if there was nothing below characters, then characters would be a "good" tag type. (Or at least, probably better than fandoms.)

Maybe it would be better to compute a goodness "score," and pick the tag type with the best score? Something like ... prompts_with_any + prompts.size * (estimated_tags_per_prompt)^2 / estimated_unique_tag_count, to get a rough estimate of the number of prompts that would be considered "candidates" when indexing on that tag type? (Ignoring the division between requests and offers for simplicity's sake.) Then pick the tag type with the lowest score. That would mean that the character tag type would pretty much always score better than the fandom type on a single-fandom exchange, even with oodles of people offering any.

As you say, though, it might be easier to tune the heuristic on real-world data.